### PR TITLE
feat(callbacks): add WandbForesightMetricsLogger

### PR DESCRIPTION
**Pull request by felixmaximilian**
## Summary

- Adds a validation-time PL callback that turns the offline foresight probe into a live training signal, logging two scalars per epoch per image source.
- `foresight/<src>/r2_hot`: linear-probe R²(pred → target) on the top-5% highest-motion patches at the deepest FD position (predicting img@T-1). Drops if the model collapses to autoencoding.
- `foresight/<src>/search_hit_hot`: cross-patch search accuracy on the same patches (chance ~1/256); complementary to R² — only satisfied when pred commits to absolute spatial positions, not just linear directions.
- Motion stratification uses the temporal delta between `last_targets[..., -1]` and `last_targets[..., -2]`, so no extra forward pass is needed. R² impl mirrors klindtlab/lejepa-identifiability's `bidirectional_r2` (in-sample lstsq, scalar SS form). Cap `max_hot_pairs=50000` bounds CPU memory and keeps n/d ≈ 130 — well above the rule-of-thumb threshold for in-sample R² discrimination.

## Context

These metrics fall out of the foresight investigation (see https://app.notion.com/p/Foresight-in-the-rmind-Control-Transformer-Does-the-Unsupervised-Last-Position-Predict-the-Future-37bd658ccf8780a3b298cc7c45304af2). The offline probe established that this pair of numbers, computed at the last FD position on the hot stratum, ranks two pretrained checkpoints consistently with their val FD loss while exposing failure modes the loss itself can't see (autoencoding collapse → R² stays high, search collapses; uniform / mean prediction → both collapse). Wiring them into validation lets training runs surface the same signal automatically.

Registered next to `WandbPatchSimilarityLogger` in `config/trainer/callbacks/pretrain.yaml`; the existing patch-similarity viz is complementary (qualitative anchor inspection on moving objects) and stays in place.

## Test plan

- [x] Smoke test: synthetic batches (signal vs. random) — signal R² ≈ 0.92 / search 1.0; random R² ≈ 0.012 / search ≈ 1/256.
- [x] Pre-commit (ruff check, ruff format, ty, deptry, typos) passes.
- [x] Run a short pretrain validation with this callback enabled and confirm wandb metrics appear under the expected keys. Use only a debugging run (only a few batches but in a way val metrics show up)

### CODE REVIEW

**Comment by felixmaximilian** [comment]
For action item 3 on the test plan:
Debug training ran locally:
Both metrics were logged to wandb under the exact expected keys during the validation epoch:
- foresight/cam_front_left/r2_hot = 0.4196
- foresight/cam_front_left/search_hit_hot = 0.004167 (≈ 1/256 chance, as expected for a randomly-initialized model)

File: src/rmind/callbacks/loggers/foresight_metrics.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+- ``{key}/{src}/r2_hot`` -- linear-probe R²(pred → target) on the
+  top-{hot_quantile} highest-motion patches at the deepest FD position
+  (predicting img@T-1). Drops if the model collapses to autoencoding.
+
+- ``{key}/{src}/search_hit_hot`` -- cross-patch search accuracy on the
+  same hot patches: for each pred[p], fraction of the time
+  ``argmin_q ||pred[p] - target[q]||²`` equals ``p`` (chance ~1/256).
+  Complementary to R²; satisfied only by absolute spatial commitment.
+
+Motivation and definitions: see ``experiments/report_short.tex``
```
Comment (naidenovaleksei) [inline src/rmind/callbacks/loggers/foresight_metrics.py]: absent link might be confusing

File: src/rmind/callbacks/loggers/foresight_metrics.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+from pydantic import validate_call
+from pytorch_lightning.callbacks import Callback
+from torch import Tensor
+from torch.utils._pytree import MappingKey, key_get, tree_map  # noqa: PLC2701
+
+from .common import _get_wandb_loggers
+
+
+@final
+class WandbForesightMetricsLogger(Callback):
```
Comment (koritsky) [inline src/rmind/callbacks/loggers/foresight_metrics.py]: i'd use `SafeCallback` here

File: src/rmind/callbacks/loggers/foresight_metrics.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            )
+            pl_module.log(
+                f"{self._key}/{src}/r2_hot", r2, sync_dist=False, rank_zero_only=True
+            )
+            pl_module.log(
+                f"{self._key}/{src}/search_hit_hot",
+                search_acc,
+                sync_dist=False,
+                rank_zero_only=True,
+            )
```
Comment (koritsky) [inline src/rmind/callbacks/loggers/foresight_metrics.py]: logging `float("nan")` to wandb feels off

```suggestion
                else float("nan")
            )
            if not math.isnan(r2):
                  pl_module.log(
                      f"{self._key}/{src}/r2_hot", r2, sync_dist=False, rank_zero_only=True
                  )
            if not math.isnan(search_hit_hot):
                    pl_module.log(
                        f"{self._key}/{src}/search_hit_hot",
                        search_acc,
                        sync_dist=False,
                        rank_zero_only=True,
                    )
```

and `import math`

**Review by koritsky — CHANGES_REQUESTED** [review]
maybe a small unit test is needed as well?

Otherwise good pr

**Comment by sandhawalia** [comment]
Could we see some visuals how does this metric look like in practice or a sanity run ?

**Comment by koritsky** [comment]
> Could we see some visuals how does this metric look like in practice or a sanity run ?

This is the comparison of given metrics for 256 vs 64 foresight tokens. don't know for now if we should reduce to 64. FT doesnt seem to suffer tho
image

File: src/rmind/callbacks/loggers/foresight_metrics.py (hunk 3 above)
Comment (felixmaximilian) [reply src/rmind/callbacks/loggers/foresight_metrics.py]: Agreed, guarded both `pl_module.log` calls with `math.isnan(...)` (+ `import math`). Heads-up: the suggested snippet checked `search_hit_hot`, but the variable in scope is `search_acc` — `search_hit_hot` is only the metric key — so pasting it verbatim would `NameError`. I guarded the actual `r2` / `search_acc`. c95e092.

File: src/rmind/callbacks/loggers/foresight_metrics.py (hunk 1 above)
Comment (felixmaximilian) [reply src/rmind/callbacks/loggers/foresight_metrics.py]: Good catch — those experiments/*.tex / *.py paths do not exist in this repo. Replaced them with the Notion investigation doc (and repointed a residual "per the report" mention in the Notes block). Fixed in c95e092.

File: src/rmind/callbacks/loggers/foresight_metrics.py (hunk 2 above)
Comment (felixmaximilian) [reply src/rmind/callbacks/loggers/foresight_metrics.py]: Done — now extends SafeCallback with fail_gracefully/disable_on_error, and the three validation hooks route through _safe_call(...) (same pattern as WandbAttentionMaskLogger). c95e092.

**Comment by felixmaximilian** [comment]
Pushed c95e092 addressing the review: added unit tests (signal / noise / NaN-skip), switched to SafeCallback, and guarded the NaN logging. Also cleaned up the dead experiments/ doc links per @naidenovaleksei. PTAL @koritsky

**Comment by felixmaximilian** [comment]
> Could we see some visuals how does this metric look like in practice or a sanity run ?

@koritsky Did you get your results via a run that's on wandb? Could you link it here in case this is true?

**Comment by koritsky** [comment]
>  > Could we see some visuals how does this metric look like in practice or a sanity run ?

> @koritsky Did you get your results via a run that's on wandb? Could you link it here in case this is true?

nope, i just recreated metrics

**Review by koritsky — APPROVED** [review]
